instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/snippet/SnippetInjectingResponseWrapper.java

```diff
@@ -0,0 +1,159 @@
+        System.err.println("Invalid string format");
+      }
+    }
+    super.setHeader(name, value);
+  }
+
+  @Override
+  public void addHeader(String name, String value) {
+    if (shouldInject() && "Content-Length".equalsIgnoreCase(name)) {
+      try {
+        value = Integer.toString(SNIPPET_LENGTH + Integer.valueOf(value));
+      } catch (NumberFormatException ex) {
+        System.err.println("Invalid string format");
+      }
+    }
+    super.addHeader(name, value);
+  }
+
+  @Override
+  public void setIntHeader(String name, int value) {
+    if (shouldInject() && "Content-Length".equalsIgnoreCase(name)) {
+      value = SNIPPET_LENGTH + value;
+    }
+    super.setIntHeader(name, value);
+  }
+
+  @Override
+  public void addIntHeader(String name, int value) {
+    if (shouldInject() && "Content-Length".equalsIgnoreCase(name)) {
+      value = SNIPPET_LENGTH + value;
+    }
+    super.addIntHeader(name, value);
+  }
+
+  @Override
+  public void setContentLength(int len) {
+    if (shouldInject()) {
+      len = len + SNIPPET_LENGTH;
+    }
+    super.setContentLength(len);
+  }
+
+  @Nullable private static final MethodHandle setContentLengthLongHandler = getMethodHandle();
+
+  @Nullable
+  private static MethodHandle getMethodHandle() {
+    try {
+      return MethodHandles.lookup()
+          .findSpecial(
+              HttpServletResponseWrapper.class,
+              "setContentLengthLong",
+              MethodType.methodType(void.class),
+              SnippetInjectingResponseWrapper.class);
+    } catch (NoSuchMethodException | IllegalAccessException e) {
+      System.out.println("SnippetInjectingResponseWrapper setContentLengthLong e: " + e);
+      return null;
+    }
+  }
+
+  private boolean shouldInject() {
+    boolean shouldInject = true;
+    if (!super.containsHeader("content-type")) {
+      shouldInject = false;
+    }
+    String contentType = super.getContentType();
+    if (contentType == null || !contentType.contains("text/html")) {
+      shouldInject = false;
+    }
+    if (super.getStatus() != 200) {
+      shouldInject = false;
+    }
+    return shouldInject;
+  }
+
+  public void setContentLengthLong(long length) throws Throwable {
+    if (shouldInject()) {
+      length = length + SNIPPET_LENGTH;
+    }
+    if (setContentLengthLongHandler == null) {
+      System.out.println("didn't find setContentLengthLong function");
+      super.setContentLength((int) length);
+    } else {
+      setContentLengthLongHandler.invokeWithArguments(this, length);
+    }
+  }
+
+  public String getCharacterEncodingHelper() {
+    String characterEncoding = super.getCharacterEncoding();
+    if (characterEncoding == null) {
+      characterEncoding = "UTF-8";
+    }
+    return characterEncoding;
+  }
+
+  @Override
+  public ServletOutputStream getOutputStream() throws IOException {
+    ServletOutputStream output = super.getOutputStream();
+    InjectionState obj = getInjectionObject(output);
```

Comment:
and in other places call `getInjectionObject` (only creating here)

then we don't need default InjectionState constructor

and also we can make characterEncoding field `final`

```suggestion
    InjectionState obj = getOrCreateInjectionObject(output, getCharacterEncodingHelper());
```